testutils/testtracing/propagation.go

```diff
@@ -0,0 +1,313 @@
+		resetSpans:    func() {},
+		isFake:        true,
+	})
+}
+
+func (s *PropagationTestSuite) runWithMockTracer(t *testing.T) {
+	mockTracer := mocktracer.New()
+	s.runWithTracer(t, tracerChoice{
+		tracerType: Mock,
+		tracer:     mockTracer,
+		spansRecorded: func() int {
+			return len(MockTracerSampledSpans(mockTracer))
+		},
+		resetSpans: func() {
+			mockTracer.Reset()
+		},
+	})
+}
+
+func (s *PropagationTestSuite) runWithJaegerTracer(t *testing.T) {
+	jaegerReporter := jaeger.NewInMemoryReporter()
+	jaegerTracer, jaegerCloser := jaeger.NewTracer(testutils.DefaultServerName,
+		jaeger.NewConstSampler(true),
+		jaegerReporter)
+	// To enable logging, use composite reporter:
+	// jaeger.NewCompositeReporter(jaegerReporter, jaeger.NewLoggingReporter(jaeger.StdLogger)))
+	defer jaegerCloser.Close()
+	s.runWithTracer(t, tracerChoice{
+		tracerType: Jaeger,
+		tracer:     jaegerTracer,
+		spansRecorded: func() int {
+			return len(jaegerReporter.GetSpans())
+		},
+		resetSpans: func() {
+			jaegerReporter.Reset()
+		},
+		zipkinCompatible: true,
+	})
+}
+
+func (s *PropagationTestSuite) runWithTracer(t *testing.T, tracer tracerChoice) {
+	testCases, ok := s.TestCases[tracer.tracerType]
+	if !ok {
+		t.Logf("No test cases for encoding=%s and tracer=%s", s.Encoding.Format, tracer.tracerType)
+		return
+	}
+	opts := &testutils.ChannelOpts{
+		ChannelOptions: tchannel.ChannelOptions{Tracer: tracer.tracer},
+		DisableRelay:   true,
+	}
+	ch := testutils.NewServer(t, opts)
+	defer ch.Close()
+	ch.Peers().Add(ch.PeerInfo().HostPort)
+	call := s.Register(t, ch)
+	for _, tt := range testCases {
+		s.runTestCase(t, tracer, ch, tt, call)
+	}
+}
+
+func (s *PropagationTestSuite) runTestCase(
+	t *testing.T,
+	tracer tracerChoice,
+	ch *tchannel.Channel,
+	test PropagationTestCase,
+	call TracingCall,
+) {
+	descr := fmt.Sprintf("test %+v with tracer %+v", test, tracer)
+	ch.Logger().Debug("")
+	ch.Logger().Debugf("======> STARTING %s", descr)
+
+	tracer.resetSpans()
+
+	span := ch.Tracer().StartSpan("client")
+	span.Context().SetBaggageItem(BaggageKey, BaggageValue)
+	ctx := opentracing.ContextWithSpan(context.Background(), span)
+
+	ctxBuilder := tchannel.NewContextBuilder(5 * time.Second).SetParentContext(ctx)
+	if test.TracingDisabled {
+		ctxBuilder.DisableTracing()
+	}
+	ctx, cancel := ctxBuilder.Build()
+	defer cancel()
+
+	req := &TracingRequest{ForwardCount: test.ForwardCount}
+	ch.Logger().Infof("Sending tracing request %+v", req)
+	response, err := call(ctx, req)
+	require.NoError(t, err)
+	ch.Logger().Infof("Received tracing response %+v", response)
+
+	// Spans are finished in inbound.doneSending() or outbound.doneReading(),
+	// which are called on different go-routines and may execute *after* the
+	// response has been received by the client. Give them a chance to run.
+	for i := 0; i < 100; i++ {
+		if spanCount := tracer.spansRecorded(); spanCount == test.ExpectedSpanCount {
+			break
+		}
+		time.Sleep(time.Millisecond) // max wait: 100ms
```

Comment:
use `testutils.Timeout(time.Millisecond)` since the Travis environment can be a lot slower, so we scale timeouts there.